src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
+        collector: Optional[CoverageCollector] = None,
     ) -> None:
         super().__init__(conjecturedata)
-        self.choices = choices
-        self.index = 0
+        self.collector = collector
+        self._index = 0
+        self._replay_queue: SortedList[ReplayQueueElement] = SortedList(
+            key=lambda x: (x[0], _choices_size(x[1]))
+        )
+
+        # TODO: thread through a random instance from FuzzProcess here, or access
+        # the current BuildContext random?
+        self.random = Random()
+        self.ninputs = 0
+        self.elapsed_time = 0.0
+        self.since_new_branch = 0
+        self.phase: Optional[Phase] = None
+        self.status_counts = StatusCounts()
+
+        self._last_report: Report | None = None
+        self._last_saved_report_at = -math.inf
+        self._last_observed = -math.inf
+        self._started = False
+
+        self.worker_identity: Optional[WorkerIdentity] = None
+        self.database_key: Optional[bytes] = None
+        self.corpus: Optional[Corpus] = None
+        self.db: Optional[HypofuzzDatabase] = None
+        # per-test-case state, reset at the beginning of each test case.
+        self._state: Optional[State] = None
+
+    def _startup(self) -> None:
+        assert (
+            settings().database is not None
+        ), "pointless to use HypofuzzProvider without a database"
+
+        # TODO: make this a context manager we enter in per_test_case_context_manager,
+        # so it resets after using hypofuzz?
+        hypothesis.internal.observability.OBSERVABILITY_CHOICES = True
+        hypothesis.internal.observability.OBSERVABILITY_COLLECT_COVERAGE = False
+
+        self.db = HypofuzzDatabase(BackgroundWriteDatabase(settings().database))
+        assert not self._started
+        context = current_build_context()
+
+        self.database_key = function_digest(context.wrapped_test.hypothesis.inner_test)  # type: ignore
+        # TODO this means our nodeid might be different in the
+        # @settings(backend="hypofuzz") case (which uses __func__.__name__)
+        # and the hypofuzz worker case (which sets the current pytest item and
+        # uses the nodeid). Is that ok? It means we can't rely on there being
+        # one nodeid across all reports for a database_key. Probably fine, just
+        # take the latest for display?
+        self.nodeid = getattr(
+            current_pytest_item.value, "nodeid", None
+        ) or get_pretty_function_description(context.wrapped_test)
+
+        self.worker_identity = worker_identity(
+            in_directory=Path(inspect.getfile(context.wrapped_test)).parent
+        )
+        self.corpus = Corpus(self.db, self.database_key)
+
+        # Report that we've started this fuzz target
+        self.db.save(test_keys_key, self.database_key)
+        # save the worker identity once at startup
+        self.db.save_worker_identity(self.database_key, self.worker_identity)
```

Comment:
I'd move these to after all the other database-loading, albeit still before we save the initial-report: if we crash before actually doing anything, cleaner if it looks like this run never occured.